lib/vmpooler/vsphere_helper.rb

```diff
@@ -196,31 +196,62 @@ def find_folder(foldername)
       base
     end
 
+    def get_host_utilization(host, model=nil, limit=90)
+      if model and not host.hardware.cpuPkg[0].description.split().include? model
+        return nil
+      end
+      if not host.runtime.inMaintenanceMode
+        if host.overallStatus == 'green'
+          cpu_usage = host.summary.quickStats.overallCpuUsage
+          cpu_size = host.summary.hardware.cpuMhz * host.summary.hardware.numCpuCores
+          cpu_utilization = (cpu_usage.to_f / cpu_size.to_f) * 100
+          memory_usage = host.summary.quickStats.overallMemoryUsage
+          memory_size = host.summary.hardware.memorySize / 1024 / 1024
+          memory_utilization = (memory_usage.to_f / memory_size.to_f) * 100
+          if not cpu_utilization > limit and not memory_utilization > limit
+            utilization = cpu_utilization + memory_utilization
+            return [ utilization, host ]
+          end
+        end
+      end
+      return nil
+    end
+
     def find_least_used_host(cluster)
       begin
         @connection.serviceInstance.CurrentTime
       rescue
         initialize
       end
 
-      hosts = {}
-      hosts_sort = {}
-
+      target_hosts = []
       datacenter = @connection.serviceInstance.find_datacenter
       datacenter.hostFolder.children.each do |folder|
         next unless folder.name == cluster
         folder.host.each do |host|
-          if
-            (host.overallStatus == 'green') &&
-            (!host.runtime.inMaintenanceMode)
-
-            hosts[host.name] = host
-            hosts_sort[host.name] = host.vm.length
-          end
+          host_usage = get_host_utilization(host)
+          target_hosts << host_usage if host_usage
         end
       end
+      return target_hosts.sort[0][1]
+    end
 
-      hosts[hosts_sort.sort_by { |_k, v| v }[0][0]]
+    def find_least_used_compatible_host(vm)
+      begin
+        @connection.serviceInstance.CurrentTime
+      rescue
+        initialize
+      end
+
+      source_host = vm.summary.runtime.host
+      model = source_host.hardware.cpuPkg[0].description.split()[4]
+      cluster = source_host.parent
+      target_hosts = []
+      cluster.host.each do |host|
+        host_usage = get_host_utilization(host, model)
+        target_hosts << host_usage if host_usage
+      end
+      return target_hosts.sort[0][1]
```

Comment:
Nitpick: we don't need `return` at the end of a method in ruby, it's enough just to say `target_hosts.sort[0][1]`.

Also, I notice that we're doing similar things here as we are higher up in the diff.  Makes me think there's a more fundamental operation happening that we could extract to a method and share?